### ISSUE TITLE
Enterprize github support?

### ISSUE THREAD

**Issue by hjri** [ralphbean/bugwarrior#360]
Currently only github service available is github.com, but if one'd like to use their company's internal github service it wouldn't be possible.

**Comment by irl** [comment]
It looks that GitHub enterprise supports the same APIs as github.com and all that would be needed is to change the API endpoint used in the existing GitHub service.

https://developer.github.com/v3/enterprise/

**Comment by irl** [comment]
@hjri Do you know of any way that testing can be performed for this without going and paying for a GitHub Enterprise subscription?

**Comment by hjri** [comment]
None that I know. I guess only way is to use user-based testing. I have access to one so i could test it a bit.

**Comment by irl** [comment]
Our general rule here is that we won't write a backend for a service unless we use it personally (it's otherwise hard to be sure that it really works). However, we gladly accept new maintainers for new backends. Would you be interested in doing the work to introduce and maintain this feature?

http://bugwarrior.readthedocs.io/en/latest/contributing.html

**Comment by hjri** [comment]
Thanks for the offer, but I'm not good at Python. I could sit for several evenings and make a PR for this issue, but I doubt code quality would be any good.

**Comment by irl** [comment]
@hjri Ok, we can leave the issue open and hopefully there would be someone who can make this happen. It really doesn't look like it would be too difficult but without someone using this who is happy to maintain it and debug any issues, it could become problematic later.

**Comment by gdetrez** [comment]
@hjri I tryed to work on this but I don't have access to a private github so I have no idea if it works or not.  I'd be happy if you could test it for me if you can.

You can install bugwarrior from [this branch](https://github.com/gdetrez/bugwarrior/tree/360-enterprize-github-support) (I suggest to do in in a virtualenv and to **backup your tasks first**):
```sh
$ pip install git+https://github.com/gdetrez/bugwarrior.git@360-enterprize-github-support 
```

Then there is a new option `github.host` where you should be able to put the hostname of your local github:
```ini
[workproject]
service = gihtub
github.host = github.acme.biz
github.login = john
...
```

**Comment by hjri** [comment]
Seems to be kinda working, after screwing around and reading manual i got it to import PRs and issues, but I get this error after running `bugwarrior-pull`
```
CRITICAL:bugwarrior.services.github: Unrecognized issue URL: https://github.acme.biz/org/repo/pull/1337.
```
(domain/organization/repository/pr number have been replaced)

**Comment by gdetrez** [comment]
Hi @hjri. I think I found the problem. I pushed new commit to the same branch, if you can git it a try.
Note that I also rebased on the latest commit from mainline so there might be unrelated issues.

**Comment by hjri** [comment]
@gdetrez doesn't seem to work
```
INFO:bugwarrior.db:Service-defined UDAs exist: you can optionally use the `bugwarrior-uda` command to export a list of UDAs you can add to your taskrc file.
INFO:bugwarrior.services:Starting to aggregate remote issues.
INFO:bugwarrior.services:Spawning 1 workers.
INFO:bugwarrior.services:Working on [work]
ERROR:bugwarrior.services:Worker for [work] failed: 'GithubService' object has no attribute 'get_repository_from_issue_url'
Traceback (most recent call last):
  File "/home/hjkos/.local/lib/python3.5/site-packages/bugwarrior/services/__init__.py", line 528, in _aggregate_issues
    for issue in service.issues():
  File "/home/hjkos/.local/lib/python3.5/site-packages/bugwarrior/services/github.py", line 386, in issues
    self.get_involved_issues(self.username).items())
  File "/home/hjkos/.local/lib/python3.5/site-packages/bugwarrior/services/github.py", line 298, in get_involved_issues
    repo = self.get_repository_from_issue_url(issue)
AttributeError: 'GithubService' object has no attribute 'get_repository_from_issue_url'
INFO:bugwarrior.services:Done with [work] in 1.124517s
INFO:bugwarrior.services:Terminating workers
CRITICAL:bugwarrior.command:Aborted (critical error in target 'work')
```

Looking over at code it seems that there's some typo - `get_repository_from_issue` exists in `GithubService` but `get_repository_from_issue_url` does not.

P.S. disregard python3.5 - on 2.7 same thing occurs, i was testing if it works on 3

**Comment by gdetrez** [comment]
Your're right, that's a typo. I pushed a new commit to fix it.

**Comment by hjri** [comment]
```
Traceback (most recent call last):
  File "/home/hjkos/.local/lib/python2.7/site-packages/bugwarrior/services/__init__.py", line 528, in _aggregate_issues
    for issue in service.issues():
  File "/home/hjkos/.local/lib/python2.7/site-packages/bugwarrior/services/github.py", line 386, in issues
    self.get_involved_issues(self.username).items())
  File "/home/hjkos/.local/lib/python2.7/site-packages/bugwarrior/services/github.py", line 298, in get_involved_issues
    repo = self.get_repository_from_issue(issue)
  File "/home/hjkos/.local/lib/python2.7/site-packages/bugwarrior/services/github.py", line 313, in get_repository_from_issue
    url = issue['repos_url']
KeyError: 'repos_url'
```

**Comment by gdetrez** [comment]
It looks like github returns different JSON objects depending on how the issue is accessed :disappointed: 
I pushed a new commit that should fix this problem (based on github API doc).

**Comment by hjri** [comment]
now it works, thanks

**Comment by gdetrez** [comment]
@hjri Cool that it finally works :smile:
As soon as I understand how to fix travis is my branches, I'll open a PR to have this merged in `develop`.

**Comment by ryneeverett** [comment]
Resolved via #360.
